### PR TITLE
[SPIR-V] Fixup storage class for global private

### PR DESCRIPTION
**Pull request by Keenuts**
Adds a new address spaces: `hlsl_private`. Variables with such address space will be emitted with a `Private` storage class.
This is useful for variables global to a SPIR-V module, since up to now, they were still emitted with a `Function` storage class, which is wrong.

### CODE REVIEW

File: llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp [hunk 1]
```diff
@@ -3337,11 +3343,8 @@ bool SPIRVInstructionSelector::selectGlobalValue(
   if (HasInit && !Init)
     return true;
 
-  unsigned AddrSpace = GV->getAddressSpace();
-  SPIRV::StorageClass::StorageClass Storage =
-      addressSpaceToStorageClass(AddrSpace, STI);
   bool HasLnkTy = GV->getLinkage() != GlobalValue::InternalLinkage &&
-                  Storage != SPIRV::StorageClass::Function;
+                  StorageClass != SPIRV::StorageClass::Function;
```
Comment (s-perron) [inline llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:3347]: Can `StorageClass` be `Function` anymore? You change it to `Static` at line 3272, right?
```suggestion
                  StorageClass != SPIRV::StorageClass::Static;
```

File: llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp [hunk 2]
```diff
@@ -3264,9 +3264,15 @@ bool SPIRVInstructionSelector::selectGlobalValue(
...
   }
-  SPIRVType *ResType = GR.getOrCreateSPIRVPointerType(
-      PointerBaseType, I, TII,
-      addressSpaceToStorageClass(GV->getAddressSpace(), STI));
+
+  const unsigned AddrSpace = GV->getAddressSpace();
+  SPIRV::StorageClass::StorageClass StorageClass =
+      addressSpaceToStorageClass(AddrSpace, STI);
+  if (StorageClass == SPIRV::StorageClass::Function)
+    StorageClass = SPIRV::StorageClass::Private;
```
Comment (s-perron) [inline llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:3272]: This is the tricky part. In some ways it makes sense. A global value then that is in `Function` scope is meaningless. The choice are that we assert or change the storage class. I don't know how that will affect the code gen for functions.

If we do change the storage class to private, is there any documentation that will need to be updated?

Could this adhoc change here cause problems if another locations calls `addressSpaceToStorageClass` to get the storage class? If they don't do the same switch could we get inconsistent results. For example, a function that returns the pointer to a variable? At the call site, they will only see an addressspace of 0. How will it know if it actually returns value is a global value (Private) or something else (Function). I'm just thinking off the top of my head. This needs to be carefully thought out.

File: llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp (hunk 1 above)
Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp]: That's correct, removed this dead code

Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp]: Wait, static?

Comment (s-perron) [reply llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp]: You were too fast. I changed it to Private after. My mistake.

File: llvm/lib/Target/SPIRV/SPIRVUtils.cpp [hunk 3]
```diff
@@ -205,6 +205,10 @@ addressSpaceToStorageClass(unsigned AddrSpace, const SPIRVSubtarget &STI) {
                : SPIRV::StorageClass::CrossWorkgroup;
   case 7:
     return SPIRV::StorageClass::Input;
+  case 8:
+    return SPIRV::StorageClass::Output;
+  case 9:
+    return SPIRV::StorageClass::Private;
```
Comment (MrSidims) [inline llvm/lib/Target/SPIRV/SPIRVUtils.cpp:211]: There are some potential synchronization issues between the repositories that we should address in the future. Would you mind to move Private to address space 10? The reason I'm asking for this is that the [llvm-spirv translator uses it for CodeSectionINTEL storage class](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/88b18319c943e70c2475b495806226e117fdaa15/lib/SPIRV/SPIRVInternal.h#L192) :)

File: llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp (hunk 2 above)
Comment (MrSidims) [reply llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:3272]: Is there a way to adjust frontend to emit the proper address space for private variables (outside of function scope) before going to the backend? It would be the cleanest solution.

File: llvm/test/CodeGen/SPIRV/extensions/SPV_INTEL_function_pointers/fp_const.ll [hunk 4]
```diff
@@ -11,7 +11,11 @@
 ; CHECK-DAG: %[[TyFun:.*]] = OpTypeFunction %[[TyInt64]] %[[TyInt64]]
 ; CHECK-DAG: %[[TyInt8:.*]] = OpTypeInt 8 0
 ; CHECK-DAG: %[[TyPtrFun:.*]] = OpTypePointer Function %[[TyFun]]
-; CHECK-DAG: %[[ConstFunFp:.*]] = OpConstantFunctionPointerINTEL %[[TyPtrFun]] %[[DefFunFp:.*]]
+
+; FIXME: OpConstantFunctionPointerINTEL requires a CodeSectionINTEL ptr.
```
Comment (MrSidims) [inline llvm/test/CodeGen/SPIRV/extensions/SPV_INTEL_function_pointers/fp_const.ll:15]: I'm fine with breaking it until we fix it by introducing CodeSectionINTEL storage class in SPIR-V backend.

File: llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp (hunk 2 above)
Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:3272]: This is indeed possible (initial PR added a `vulkan_private` address space).
But if the proposed RFC to set AS0=generic doesn't come true, I'd tend to believe AS0+global should end up in the private class, like AS0+local ends up in the Function class.

File: llvm/lib/Target/SPIRV/SPIRVUtils.cpp (hunk 3 above)
Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVUtils.cpp:211]: Sure

File: llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp (hunk 2 above)
Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:3272]: I agree with concerns stated above by @s-perron , and as a part of instruction selection this substitution potentially makes the code more fragile and prone to error. The substitution itself however makes sense, so I think we can just move it to a later stage, when SPIR-V code is (almost) generated. Probably a more safe place for this substitution is SPIRVTargetLowering::finalizeLowering() where we keep all post-generation validation actions.

Additional benefit would be that all these kind of things are concentrated in one place, making post-generation validation a bit self-documented, and since this substitution logically belongs to the same sort of actions, its presence in SPIRVTargetLowering::finalizeLowering() among similar cases is more obvious and less unexpected for a developer.

Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:3272]: Actually, there is even better and simpler approach, similar to what @MrSidims asked about adjusting a frontend. We can do the substitution much much earlier, *before* IR Translation, to be sure that this is not intertwined with type inference and duplicates tracker.

Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:3272]: E.g., something along the lines of

```
bool SPIRVEmitIntrinsics::runOnModule(Module &M) {
  bool Changed = false;

  for (auto GV = M.global_begin(), GE = M.global_end(); GV != GE; ++GV) {
    if (isa<Function>(GV) || GV->getAddressSpace() != storageClassToAddressSpace(SPIRV::StorageClass::Function))
      continue;
    // add to worklist
  }
  // traverse worklist, re-create the global in another address space GV->replaceAllUsesWith(newGV);
```

**Comment by Keenuts** [comment]
Hello all!
Changed the PR to use a new AS emitted by the FE. This way there are no more weird storage class switch.
However, this required to support a new `addrspacecast` operation from `Function` to `Private` and the other way around.

The FE change will be in another PR (can be merged later on)

**Comment by Keenuts** [comment]
@llvm-beanz : FYI for the added address space.

**Review by s-perron — COMMENTED** [review]
The implementation looks good to me. I just wonder about one of the tests. I'd like to know why we need to handle that case, and how prevalent the `addrspacecast`s will be.

File: llvm/test/CodeGen/SPIRV/pointers/global-addrspacecast.ll [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+; RUN: %if spirv-tools %{ llc -O0 -mtriple=spirv64-unknown-unknown %s -o - -filetype=obj | spirv-val %}
+
+@PrivInternal = internal addrspace(10) global i32 456
+; CHECK-DAG:  %[[#type:]] = OpTypeInt 32 0
+; CHECK-DAG: %[[#ptrty:]] = OpTypePointer Private %[[#type]]
+; CHECK-DAG: %[[#value:]] = OpConstant %[[#type]] 456
+; CHECK-DAG:   %[[#var:]] = OpVariable %[[#ptrty]] Private %[[#value]]
+
+define spir_kernel void @Foo() {
+  %p = addrspacecast ptr addrspace(10) @PrivInternal to ptr
```
[the remark this answers is not in the document]
Comment (Keenuts) [reply llvm/test/CodeGen/SPIRV/pointers/global-addrspacecast.ll:11]: This test is specifically to validate we can lower an addrspacecast from private to function.

This pattern can be generated by the actual codegen:
- https://clang.llvm.org/doxygen/CodeGenModule_8cpp_source.html#l05259
- https://clang.llvm.org/doxygen/CodeGenModule_8cpp_source.html#l04982

What happens is a global is recreated with the correct AS, and where a load was done (or function call with this as parameter), the `addrspacecast` is added.
In basic cases, I don't think this is an issue, as it generates a `load(addrspacecast(ptr))`.

BUT I suspect if we start to have references, then it could become an issue, as the ptr will be changed to another AS, and suddenly the `OpTypePtr` becomes `Function` instead of `Private`?

File: llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp [hunk 6]
```diff
@@ -3388,6 +3398,13 @@ bool SPIRVInstructionSelector::selectGlobalValue(
         GVType, MIRBuilder, SPIRV::AccessQualifier::ReadWrite, false);
   }
 
+  const unsigned AddrSpace = GV->getAddressSpace();
+  SPIRV::StorageClass::StorageClass StorageClass =
+      addressSpaceToStorageClass(AddrSpace, STI);
+
+  SPIRVType *ResType =
+      GR.getOrCreateSPIRVPointerType(PointerBaseType, I, TII, StorageClass);
+
```
[the remark this answers is not in the document]
Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:3407]: Indeed, Just checked `ResType` usage, but turns out it's redefined in the condition scope. Moved all this code down.

**Review by MrSidims — APPROVED** [review]
LGTM
Guess the PR description should be updated removing CodeSectionINTEL mentioning

File: llvm/test/CodeGen/SPIRV/pointers/global-addrspacecast.ll (hunk 5 above)
[the remark this answers is not in the document]
Comment (s-perron) [reply llvm/test/CodeGen/SPIRV/pointers/global-addrspacecast.ll:11]: Okay, we can leave this for now. I'll want to see how this code pattern is generated by clang/llvm because I fear it could lead to invalid spir-v, but we can deal with that when we start using the address space in clang.

**Review by s-perron — APPROVED** [review]

**Review by VyacheslavLevytskyy — APPROVED** [review]

File: llvm/test/CodeGen/SPIRV/pointers/global-addrspacecast.ll (hunk 5 above)
[the remark this answers is not in the document]
Comment (Keenuts) [reply llvm/test/CodeGen/SPIRV/pointers/global-addrspacecast.ll:11]: Yes, I actually have 1 pattern that creates a bad pattern before inline. Trying to find a good way, but haven't so far.
But that's for the FE so fine for this PR.

File: clang/test/SemaTemplate/address_space-dependent.cpp [hunk 7]
```diff
@@ -102,7 +102,7 @@ int main() {
   HasASTemplateFields<1> HASTF;
   neg<-1>(); // expected-note {{in instantiation of function template specialization 'neg<-1>' requested here}}
   correct<0x7FFFE9>();
-  tooBig<8388650>(); // expected-note {{in instantiation of function template specialization 'tooBig<8388650L>' requested here}}
+  tooBig<8388651>(); // expected-note {{in instantiation of function template specialization 'tooBig<8388651L>' requested here}}
```
Comment (arsenm) [inline clang/test/SemaTemplate/address_space-dependent.cpp:105]: Looks unrelated

File: clang/include/clang/Basic/AddressSpaces.h [hunk 8]
```diff
@@ -58,6 +58,7 @@ enum class LangAS : unsigned {
 
   // HLSL specific address spaces.
   hlsl_groupshared,
+  hlsl_private,
```
[the remark this answers is not in the document]
Comment (Keenuts) [reply clang/include/clang/Basic/AddressSpaces.h:61]: We hoped to split the PRs between back-end and FE, shall I land both at the same time?

[the remark this answers is not in the document]
Comment (arsenm) [reply clang/include/clang/Basic/AddressSpaces.h:61]: This isn't split though, the backend parts don't require any of these clang changes

[the remark this answers is not in the document]
Comment (Keenuts) [reply clang/include/clang/Basic/AddressSpaces.h:61]: Oh didn't knew I could modify the target's AS maps without also modifying this.
I'll revert/re-land this PR (given the other test), and see if I can do that:
https://github.com/llvm/llvm-project/pull/118312
